### PR TITLE
Fix recipe command on Paper

### LINKED ISSUE #1289: Duplication glitch when using /recipe
If you do /recipe diamond_block, then open your recipe book and click anything in it, it will send all the items from the workbench to your inventory.

### LINKED ISSUE #1397: When will /recipe be back?
Hi, on b499 /recipe is temporarily disabled. Was wondering why and when this will be fixed?

### PR DESCRIPTION
**Pull request by pop4959**
Based on PlayerRecipeBookClickEvent added in PaperMC/Paper/pull/3351. Fixes #1289, closes #1397.

### CODE REVIEW

**Review by JRoy — DISMISSED** [review]

File: providers/PaperProvider/src/net/ess3/provider/providers/PaperPlayerRecipeBookClickEventProvider.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package net.ess3.provider.providers;
+
+import com.destroystokyo.paper.event.player.PlayerRecipeBookClickEvent;
+import net.ess3.provider.EventProvider;
+import org.bukkit.event.Event;
+import org.bukkit.event.EventHandler;
+
+import java.util.function.Consumer;
+
+public class PaperPlayerRecipeBookClickEventProvider extends EventProvider {
```
Comment (mdcfe) [inline providers/PaperProvider/src/net/ess3/provider/providers/PaperPlayerRecipeBookClickEventProvider.java]:
```suggestion
public class PaperRecipeBookListener extends ProviderListener {
```

File: Essentials/src/com/earth2me/essentials/Essentials.java [hunk 2]
```diff
@@ -269,6 +273,18 @@ public void onEnable() {
                     serverStateProvider = new ReflServerStateProvider(getLogger());
                 }
 
+                //Event Providers
+                if (PaperLib.isPaper()) {
+                    try {
+                        Class.forName("com.destroystokyo.paper.event.player.PlayerRecipeBookClickEvent");
+                        recipeBookEventProvider = new PaperRecipeBookListener(event -> {
+                            if (this.getUser(((PlayerEvent) event).getPlayer()).isRecipeSee()) {
+                                ((Cancellable)event).setCancelled(true);
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/Essentials.java]:
```suggestion
                                ((Cancellable) event).setCancelled(true);
```

File: Essentials/src/com/earth2me/essentials/commands/Commandrecipe.java [hunk 3]
```diff
@@ -26,18 +25,16 @@
...
-        }
-    }
 
     @Override
     public void run(final Server server, final CommandSource sender, final String commandLabel, final String[] args) throws Exception {
-        disableCommandForVersion1_12();
+        try {
+            Class.forName("com.destroystokyo.paper.event.player.PlayerRecipeBookClickEvent");
+        } catch (ClassNotFoundException e) {
+            sender.sendMessage(tl("unsupportedFeature"));
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandrecipe.java]: This breaks the command on 1.11 and lower. Maybe add a 1.12 check here?

File: Essentials/src/com/earth2me/essentials/commands/Commandrecipe.java [hunk 4]
```diff
@@ -26,18 +25,16 @@
     public Commandrecipe() {
         super("recipe");
     }
-    
-    private void disableCommandForVersion1_12() throws Exception {
-        VersionUtil.BukkitVersion version = VersionUtil.getServerBukkitVersion();
-        if (version.isHigherThanOrEqualTo(VersionUtil.v1_12_0_R01)
-            && !ess.getSettings().isForceEnableRecipe()) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandrecipe.java:33]: If we're removing this check entirely, remove `Settings#isForceEnableRecipe` too (it's an undocumented config option I used to confirm whether this bug was fixed in newer versions of Spigot).

File: Essentials/src/com/earth2me/essentials/commands/Commandrecipe.java (hunk 3 above)
Comment (pop4959) [reply Essentials/src/com/earth2me/essentials/commands/Commandrecipe.java]: Class::forName doesn't work? I'll look into this.

File: Essentials/src/com/earth2me/essentials/commands/Commandrecipe.java (hunk 4 above)
Comment (pop4959) [reply Essentials/src/com/earth2me/essentials/commands/Commandrecipe.java:33]: Will do

File: Essentials/src/com/earth2me/essentials/commands/Commandrecipe.java (hunk 3 above)
Comment (mdcfe) [reply Essentials/src/com/earth2me/essentials/commands/Commandrecipe.java]: @pop4959 No, this command still works (as of the latest build) on 1.11.x and below, but by removing the 1.12 version check entirely we drop support for this command on old versions of MC. These older versions are the ones that don't have the recipe book, and so they're the versions where this command is the *most* useful.

Comment (pop4959) [reply Essentials/src/com/earth2me/essentials/commands/Commandrecipe.java]: Ah right, completely overlooked that. I'll re-add this check ASAP.

**Review by JRoy — APPROVED** [review]

**Pull request merged**